doc/user/interop.md

```diff
@@ -89,150 +90,110 @@ Returns true only for the `nil` object.
 
 ### `HAS_KEYS`
 
-Returns true for all objects except those primitives that are always frozen,
-`nil`, `true`, `false`, `Fixnum`, `Bignum`, `Float`, `Symbol`. Note that `KEYS`
-will continue to return an empty list, even if `HAS_KEYS` returns false.
+Returns true.
 
 ### `KEYS`
 
-If the receiver is a Ruby `Hash`, return the hash keys.
+If the receiver is a Ruby `Array`, return an array of indices in
+bounds.
 
-`KEYS(hash)` → `hash.keys`
+Otherwise, if the receiver is a Ruby `Hash`, return the hash keys converted to
+strings.
 
-Otherwise, return the instance variable names, without the leading `@`.
-
-`KEYS(other)` → `other.instance_variables.map { |key| ...key without @... }`
-
-In both cases if the `internal` flag is set, instance variable names with the
-leading `@` are also returned.
+Otherwise, return all method names via `receiver.methods`, and instance
+variable names if the `internal` flag is set.
 
 Keys are returned as a Ruby `Array` containing Ruby `String` objects.
 
 ### `KEY_INFO`
 
-If the object is a Ruby `Array` and the key is an integer:
+If the receiver is a Ruby `Array`:
 
-`READABLE` will be set if the index is in bounds (0 <= index < array.size)
+- `READABLE` will set if the name is an integer and in bounds.
 
-`INSERTABLE` and `MODIFIABLE` will be set if the index is in bound and the array
-is not frozen.
+- `INSERTABLE` and `MODIFIABLE` will be set if the index is an integer, in bound
+  and the array is not frozen.
 
-`REMOVABLE` will be set if the array is not frozen.
+- `REMOVABLE`, `INVOCABLE` and `INTERNAL` will not be set.
 
-If the object is a Ruby `Hash`:
+Otherwise, if the receiver is a Ruby `Hash`:
 
-`READABLE` will be set if the key is found.
+- `READABLE` will be set if the key is found.
 
-`INSERTABLE` will be set if the hash is not frozen.
+- `INSERTABLE` will be set if the hash is not frozen.
 
-`MODIFIABLE` will be set if the key is found and the hash is not frozen.
+- `MODIFIABLE` and `REMOVABLE`  will be set if the key is found and the hash is
+  not frozen.
 
-`REMOVABLE` will be set if the array is not frozen.
+- `INVOCABLE` and `INTERNAL` will not be set.
 
-Otherwise:
+Otherwise if the name starts with an `@`:
 
-`READABLE` will be set if the object responds to a method of the same name.
+- `READABLE` will be set if the instance variable exists.
 
-`MODIFIABLE` will be set if the object responds to a method of the same name
-appended with `=` and the object is not frozen.
+- `INSERTABLE` will be set if the object is not frozen.
 
-`INSERTABLE` will be false.
+- `MODIFIABLE` and `REMOVABLE` will be set if the instance variable exists is
```

Comment:
Missing commas between the conditions make this sentence hard to read.

Reply:
Fixing.